src/Sylius/Behat/Client/RequestFactory.php

```diff
@@ -0,0 +1,179 @@
+        );
+        $builder->withHeader('CONTENT_TYPE', 'application/ld+json');
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+
+    public function delete(
+        string $section,
+        string $resource,
+        string $id,
+        string $authorizationHeader,
+        ?string $token = null
+    ): RequestInterface {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s/%s', $section, $resource, $id),
+            HttpRequest::METHOD_DELETE,
+        );
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+
+    public function transition(string $section, string $resource, string $id, string $transition): RequestInterface
+    {
+        return self::customItemAction($section, $resource, $id, HttpRequest::METHOD_PATCH, $transition);
+    }
+
+    public function customItemAction(string $section, string $resource, string $id, string $type, string $action): RequestInterface
+    {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s/%s/%s', $section, $resource, $id, $action),
+            $type,
+        );
+        $builder->withHeader('CONTENT_TYPE', self::resolveHttpMethod($type));
+
+        return $builder->build();
+    }
+
+    public function upload(
+        string $section,
+        string $resource,
+        string $authorizationHeader,
+        ?string $token = null
+    ): RequestInterface {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s', $section, $resource),
+            HttpRequest::METHOD_POST,
+        );
+        $builder->withHeader('CONTENT_TYPE', 'multipart/form-data');
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+//    Something like "custom" should not exist in my opinion
+    public static function custom(string $url, string $method, array $additionalHeaders = [], ?string $token = null): RequestInterface
+    {
+        $builder = RequestBuilder::create($url, $method);
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+        $builder->withHeader('CONTENT_TYPE', self::resolveHttpMethod($method));
+
+        if ($token) {
+            $builder->withHeader('HTTP_Authorization', 'Bearer ' . $token);
+        }
+
+        foreach ($additionalHeaders as $name => $value) {
+            $builder->withHeader($name, $value);
+        }
+
+        return $builder->build();
+    }
+
+    private static function resolveHttpMethod(string $method): string
```

Comment:
If you decide to keep this class open for extending, we should make this method protected or extract it to separate service as it is used in 5 places 
```suggestion
    protected static function resolveHttpMethod(string $method): string
```

Reply:
I created ContentTypeGuide - a small little service that we can forget about ;p